src/datanode/src/heartbeat/handler/flush_region.rs

```diff
@@ -24,106 +27,210 @@ use crate::error;
-                })),
-                Err(err) => Some(InstructionReply::FlushRegion(SimpleReply {
-                    result: false,
-                    error: Some(format!("{err:?}")),
-                })),
+        }
+
+        FlushRegionReply::from_results(results)
+    }
+
+    /// Flushes a single region synchronously with proper error handling.
+    async fn flush_single_region_sync(&self, region_id: RegionId) -> Result<(), String> {
+        // Check if region is leader and writable
+        let Some(writable) = self.region_server.is_region_leader(region_id) else {
+            return Err("Region is not leader".to_string());
+        };
+
+        if !writable {
+            return Err("Region is not writable".to_string());
+        }
+
+        // Register and execute the flush task
+        let region_server_moved = self.region_server.clone();
+        let register_result = self
+            .flush_tasks
+            .try_register(
+                region_id,
+                Box::pin(async move {
+                    let request = RegionRequest::Flush(RegionFlushRequest {
+                        row_group_size: None,
+                    });
+                    region_server_moved
+                        .handle_request(region_id, request)
+                        .await?;
+                    Ok(())
+                }),
+            )
+            .await;
+
+        if register_result.is_busy() {
+            warn!("Another flush task is running for the region: {region_id}");
+        }
+
+        let mut watcher = register_result.into_watcher();
+        match self.flush_tasks.wait_until_finish(&mut watcher).await {
+            Ok(()) => Ok(()),
+            Err(err) => Err(format!("{err:?}")),
+        }
+    }
+
+    /// Enhanced handler for FlushRegionsV2 with unified semantics.
+    pub(crate) fn handle_flush_regions_v2_instruction(
+        self,
+        flush_regions_v2: FlushRegionsV2,
+    ) -> BoxFuture<'static, Option<InstructionReply>> {
+        Box::pin(async move {
+            let start_time = Instant::now();
+
+            if matches!(flush_regions_v2.strategy, FlushStrategy::Async) {
+                // Asynchronous hint mode: fire-and-forget, no reply expected
+                self.handle_flush_hint(flush_regions_v2.region_ids).await;
+                None
+            } else {
+                // Synchronous mode: return reply with results
+                let reply = self
+                    .handle_flush_sync(flush_regions_v2.region_ids, flush_regions_v2.error_strategy)
+                    .await;
+                let elapsed = start_time.elapsed();
+                debug!(
+                    "FlushRegionsV2 sync: {:?}, elapsed: {:?}, success: {}",
+                    reply.results.iter().map(|(id, _)| id).collect::<Vec<_>>(),
+                    elapsed,
+                    reply.overall_success
+                );
+                Some(InstructionReply::FlushRegionsV2(reply))
```

Comment:
Suggest moving this log out of the if-else: record for the async hint mode, too.

```suggestion
            let reply = if matches!(flush_regions_v2.strategy, FlushStrategy::Async) {
                // Asynchronous hint mode: fire-and-forget, no reply expected
                self.handle_flush_hint(flush_regions_v2.region_ids).await;
                None
            } else {
                // Synchronous mode: return reply with results
                let reply = self
                    .handle_flush_sync(flush_regions_v2.region_ids, flush_regions_v2.error_strategy)
                    .await;
                Some(InstructionReply::FlushRegionsV2(reply))
            }
            
            let elapsed = start_time.elapsed();
            debug!("{flush_regions_v2:?}, elapsed: {elapsed:?}, reply: {reply:?}");
            
            reply
```